### ISSUE TITLE
Convert holy sites

### ISSUE THREAD

**Issue by IhateTrains** [ParadoxGameConverters/ImperatorToCK3#38]
I think a hybrid solution would be nice. Mark some holy sites in blankMod as permanent (probably with a comment that the converter will recognize), validate the rest (for example check if the site was a holy site in Imperator), and in place of those that failed the validation generate new ones based on Imperator.

**Comment by RaduM84** [comment]
I suggest we seek to determine the locations of 5 Holy Sites per faith (less if there are that few provinces with that faith around).

The algorithm would look like this :
1) For each Imperator Religion.
2) Collect all Imperator Territories that have that Religion.
3) Split the Territories into 2 sets : Territories that have a Holy Site and Territories that do not.
4) Order both sets in descending order by Population.
5) Take the top 5 Territories with a Holy Site. 
6) Take the most populated Territory without a Holy Site.
7) At the end we have a list of 6 or less Imperator Territories that are marked as CK3 Holy Sites (5 or less CK3 Holy Sites that are mapped from Imperator Holy Site and one "Popular" CK3 Holy Site that is mapped from the most populated Imperator Territory without a Holy Site).
8) When mapping CK3 Baronies, if an Imperator Territory maps to multiple CK3 Baronies, choose one CK3 Barony at random.

In terms of effects, we will need to map to CK3 each Imperator Deity (I recommend only the Passive) Effect as well as the Imperator Religion Population Effect. The 5 Holy Sites or less will have CK3 Holy Site Effects that correspond to the Imperator Deity the Holy Site was dedicated to. The Holy Site that corresponds to the No-Holy Site Imperator Territory will have its effect the mapped to the population-wide Imperator Religion Effect (for example, the +8% National Citizen Happiness for Hellenism).

I volunteer to create these religious effect mappings.

**Comment by IhateTrains** [comment]
> The algorithm would look like this :
> 
> 1. For each Imperator Religion.
> 2. Collect all Imperator Territories that have that Religion.
> 3. Split the Territories into 2 sets : Territories that have a Holy Site and Territories that do not.
> 4. Order both sets in descending order by Population.
> 5. Take the top 4 Territories with a Holy Site.
> 6. Take the most populated Territory without a Holy Site.
> 7. At the end we have a list of 5 or less Imperator Territories that are marked as CK3 Holy Sites (4 or less CK3 Holy Sites that are mapped from Imperator Holy Site and one "Popular" CK3 Holy Site that is mapped from the most populated Imperator Territory without a Holy Site).
> 8. When mapping CK3 Baronies, if an Imperator Territory maps to multiple CK3 Baronies, choose one CK3 Barony at random.

I like the algorithm. The only problem I see is that religions are not always mapped 1-1 and there are regional mappings, so I would rephrase the beginning like this:

> For each CK3 religion religion:
> Collect all Imperator Territories that are mapped to this religion.

Does this make sense?

Also, as written in the original comments, I think some holy sites should be permanent (exempt from dynamic holy sites conversion), for example Christian faiths should always have Jerusalem as a holy site. The non-permanent sites could be used as fallback if we have less than 5 permanent+dynamic sites.

> In terms of effects, we will need to map to CK3 each Imperator Deity (I recommend only the Passive) Effect as well as the Imperator Religion Population Effect. The 4 Holy Sites or less will have CK3 Holy Site Effects that correspond to the Imperator Deity the Holy Site was dedicated to. The Holy Site that corresponds to the No-Holy Site Imperator Territory will have its effect the mapped to the population-wide Imperator Religion Effect (for example, the +8% National Citizen Happiness for Hellenism).
> 
> I volunteer to create these religious effect mappings.

I think it can work, but I'm not experienced enough with data files and mapping to be sure. @IohannesIohannium what do you think?
If you provide a few proof-of-concept mappings, it will probably be easier for me to tell.

**Comment by RaduM84** [comment]
No objections to the proposed algorithm revision. Also, pretty much any religion that is outside the map or outside the era should have its vanilla holy sites (so if I convert an 100 AD save from a vanilla Imperator, which allows you to play past the end date, then Christianity in that "universe" should have its historical holy sites because why not)

If Christianity exists in Imperator due to a mod... I guess your suggestion works as well as any.

I will put together some Data File examples soon.

**Comment by RaduM84** [comment]
The mapping file would be something along the lines of :

link = { imp = global_citizen_happyness ck3 = county_opinion_add factor = 100 }

where :

imp = modifier either of the religion or the passive modifier of the deity
ck3 = equivalent character modifier 
factor = floating point multiplier to the value of the Imperator modifier.

A concrete example, for Hellenism, it's Religion Modifier is Global Citizen Happiness +8%.
Using the mapping link, this transforms into CK3 as 

character_modifier = {
name = <something,something>
county_opinion_add = 8 (0.08 * 100, meaning value of imperator modifier * factor)
}

Does this look ok to you?

**Comment by RaduM84** [comment]
Regarding the on-map placement itself : 
In CK3 Holy Sites can be declared "per county" (as with most) or "per barony" such as Rome :
rome = {
	county = c_roma
	barony = b_vaticano
	
	character_modifier = {
		name = holy_site_rome_effect_name
		development_growth_factor = 0.1
		stewardship = 1
	}
}

Because in Imperator Holy Sites are built in specific Territories, and thus one can have multiple Holy Sites per province, we should reflect this by having all Imperator Holy Sites declared "per barony".

**Comment by IhateTrains** [comment]
> Because in Imperator Holy Sites are built in specific Territories, and thus one can have multiple Holy Sites per province, we should reflect this by having all Imperator Holy Sites declared "per barony".

Agreed.

**Comment by IhateTrains** [comment]
comments on loc generation https://discord.com/channels/612683871112396800/692883287336681592/972234140411899944
